### PR TITLE
:bug: token authenticate

### PR DESCRIPTION
**Pull request by roceb**
fix for issue / #619 
Missing Bearer token return error instead of panicking. 
Also updated [Client.send() method](https://github.com/konveyor/tackle2-hub/blob/main/binding/client.go#L639) to work with bearer tokens.

### CODE REVIEW

File: auth/builtin.go [hunk 1]
```diff
@@ -21,8 +22,7 @@ type Validator interface {
 }
 
 // NoAuth provider always permits access.
-type NoAuth struct {
-}
+type NoAuth struct{}
```
Comment (jortel) [inline auth/builtin.go:25]: Unnecessary reformatting not consistent with rest of the code-base.

File: auth/builtin.go [hunk 2]
```diff
@@ -59,13 +59,28 @@ func (r *NoAuth) Refresh(refresh string) (token Token, err error) {
 }
 
 // Builtin auth provider.
-type Builtin struct {
+type Builtin struct{}
```
Comment (jortel) [inline auth/builtin.go:62]: Unnecessary reformatting not consistent with rest of the code-base.

File: auth/builtin.go [hunk 3]
```diff
@@ -59,13 +59,28 @@ func (r *NoAuth) Refresh(refresh string) (token Token, err error) {
...
+
+// Parse Token out of a string
+func ParseToken(requestToken string) (string, error) {
```
Comment (jortel) [inline auth/builtin.go:65]: Why is this an exported package function instead of a non-exported method on Builtin?

File: auth/builtin.go [hunk 4]
```diff
@@ -59,13 +59,28 @@ func (r *NoAuth) Refresh(refresh string) (token Token, err error) {
...
+	return token, nil
 }
 
 // Authenticate the token
 func (r *Builtin) Authenticate(request *Request) (jwToken *jwt.Token, err error) {
-	token := strings.Replace(request.Token, "Bearer", "", 1)
-	token = strings.Fields(token)[0]
+	token, err := ParseToken(request.Token)
+	if err != nil {
+		return nil, err
```
Comment (jortel) [inline auth/builtin.go:82]: When named return values (pattern is used) the explicit assigned return values here is redundant.  `err` is already assigned and jwtToken is already nil.

Comment (jortel) [reply auth/builtin.go:82]: Since parseToken() returns an error, the call should happen after the defer so that it is logged.

File: auth/builtin.go [hunk 5]
```diff
@@ -59,13 +59,28 @@ func (r *NoAuth) Refresh(refresh string) (token Token, err error) {
...
-type Builtin struct {
+type Builtin struct{}
+
+// Parse Token out of a string
+func ParseToken(requestToken string) (string, error) {
+	splitToken := strings.Split(requestToken, " ")
+	if splitToken[0] != "Bearer" {
+		return "", errors.New("authentication header not of type bearer")
+	}
+	token := strings.TrimSpace(splitToken[1])
```
Comment (jortel) [inline auth/builtin.go:70]: This will result in a SEGV when requestToken=`Bearer`. no token.

parseToken() should return `NotValid` which will result in the request http=403 (forbidden) instead of 500.

**Review by jortel — CHANGES_REQUESTED** [review]
I do not think changes to the AUTH layer is a good first issue.

File: auth/builtin.go [hunk 6]
```diff
@@ -59,13 +59,28 @@ func (r *NoAuth) Refresh(refresh string) (token Token, err error) {
 }
 
 // Builtin auth provider.
-type Builtin struct {
+type Builtin struct{}
+
+// Parse Token out of a string
+func ParseToken(requestToken string) (string, error) {
+	splitToken := strings.Split(requestToken, " ")
```
[the remark this answers is not in the document]
Comment (roceb) [reply auth/builtin.go:66]: Bearer is for the type of [schema](https://datatracker.ietf.org/doc/html/rfc6750) and JWT is the encoding. JWT encodes to the Bearer schema. So unless there are plans to support different schemas, then Bearer should be included.

As far as:
` requestToken = strings.TrimSpace(requestToken)`
Is there a use case for " Bearer "? This seams like a bad formatted request and shouldn't be allowed.

**Comment by roceb** [comment]
We all have to start somewhere.

File: auth/builtin.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jortel) [reply auth/builtin.go]: > Bearer is for the type of [schema](https://datatracker.ietf.org/doc/html/rfc6750) and JWT is the encoding. JWT encodes to the Bearer schema. So unless there are plans to support different schemas, then Bearer should be included.

Yes, I understand the difference.

> Is there a use case for " Bearer "? This seams like a bad formatted request and shouldn't be allowed.

 Just suggesting that API could be tolerant of curl users trying to do the right thing.
That said, I suppose the standard is not that hard to follow.

**Comment by jortel** [comment]
Thanks for the PR and requested changes @roceb.

We still need to change the [Client.send() method](https://github.com/konveyor/tackle2-hub/blob/main/binding/client.go#L639) like:
```
- request.Header.Set(api.Authorization, r.token.Token)
+ request.Header.Set(api.Authorization, "Bearer "+r.token.Token)
```
Else, the binding won't work with our API.

**Comment by roceb** [comment]
Thank you for your time and patience @jortel . I have updated [Client.send() method](https://github.com/konveyor/tackle2-hub/blob/main/binding/client.go#L639). Let me know if there is anything else?

**Review by jortel — APPROVED** [review]
LGTM
